model/project.rb

```diff
@@ -57,4 +57,12 @@ def current_invoice
 
     Invoice.new(project_id: id, content: content, begin_time: begin_time, end_time: end_time, created_at: Time.now, status: "current")
   end
+
+  def get_feature_flag(name)
+    feature_flags[name]
+  end
```

Comment:
Actually, I don't want to add this. `feature_flags[name]` is shorter anyway :smile:, but we are already adding setter and adding getter only seems natural. However if reviewers agree with me, I'd happily remove this :smile:

Reply:
It might be one of the terrible ideas. What do you think about defining feature flag before use them, similar to semaphores? Freeform string doesn't feel very safe. And it's hard to remember we have lots of feature flags.

_(method names are up to you)_
```ruby
  def self.feature_flag(*flags)
    flags.map!(&:to_s)
    flags.each do |flag|
      define_method "enable_#{flag}" do |value|
        update(feature_flags: feature_flags.merge({flag => value}))
      end

      define_method "get_#{flag}" do
        feature_flags[name]
      end
    end
  end
```

Then we can define feature flags at project.rb

```ruby
    feature_flag :postgres_service, :new_image
```

Finally we can use predefined helpers

```ruby
project.enable_postgres_service(1)

project.get_postgres_service
```

Reply:
Autocompletion will work in pry.

Reply:
what about just using semaphores for a project?

Reply:
Projects don't have a strand, we can't give a valid strand_id. Also, I don't think adding extra responsibilities to the strand is a good idea.

Reply:
Also flag values are not necessarily boolean, it could be number or string, so semaphores wouldn't work there.

About meta-programming part... I'm not sure. I think the overall idea is fine but I have one reservation; name of the generated methods would be very long; `set_postgres_service` is meaningless, so we need something like `set_postgres_service_feature_flag` which is quite long. I'll think about it.

Reply:
OK, I guess it depends on how you name the feature flag. For example if you name it "enable_postgres", then set_enable_postgres is short and meaningful. So I changed the commit to use metaprogramming.

Reply:
Instead of exploding the methods into the class, you could stuff it into a "features" method, e.g. `foo.features.postgres?` (which, inside `foo` while be rendered "features.postgres?").

This may have problems that make it not so attractive, in test or something.